### PR TITLE
Graph and Partition

### PR DESCRIPTION
**Pull request by bsuwal**
Some particular things that I think might be wonky about this PR:

1. The Graph struct has a `edge_src` and a `edge_dst` array. No particular reason to store edges like this, but I thought we might as well leave a src and dst capability if we want to use this module for graphs that are directed.

2. I do not like the way Graph and Partition take in a `raw_graph`, (which is the dict I get after running JSON.parse(filepath_to_json_file) but I found it awkward to pass a filepath to both of these and make a new Graph inside Partition ( the repeated making of the Graph was awkward) so I had them take the JSON.parse - ed dicts. The utils.jl file has a get_graph_and_partition() function that I envison would be the cleanest to use so that the user doesn't have to deal with this. I don't particularly like this messiness and would love to hear thoughts.

3. The Partition function was getting bulky so I split it into the numerous functions below in that file.

4. The tests are on a toy graph I made, which are in the graph.jl and partition.jl files inside the directory `test/`. This testing standard was adopted from the LightGraphs library. https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/test/runtests.jl

### CODE REVIEW

**Comment by bsuwal** [comment]
Parker - I also know you use BitSet for some of your data structures, I dont understand them very well so would like to know if you think they would be more relevant

File: test/partition.jl [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    @test partition.district_adj[1,2] == 2
+    @test partition.district_adj[1,3] == 2
+    @test partition.district_adj[1,4] == 0
+
+    @test partition.district_nodes[1] == Set{Int}([1, 2, 5, 6])
+    @test partition.district_nodes[2] == Set{Int}([3, 4, 7, 8])
+    @test partition.district_nodes[3] == Set{Int}([9, 10, 13, 14])
+    @test partition.district_nodes[4] == Set{Int}([11, 12, 15, 16])
+
+    @test partition.assignments[1] == 1
```
Comment (pjrule) [inline test/partition.jl]: Is there any way to use a loop or series of loops here instead? I tend to favor avoiding too much abstraction in unit tests, but literally checking the assignment of _every single unit_ seems excessive! :)

File: test/graph.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
+
+
+# the test graph being loaded is labeled as
```
Comment (pjrule) [inline test/graph.jl]: For long comments, I prefer the multi-line comment style—that way, autodoc tools can pick them up.

File: src/graph.jl [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    num_nodes::Int
+    num_edges::Int
+
+    # an adjacency matrix where each entry is the edge id that connects
+    # the two nodes.
+    adj_matrix::SparseMatrixCSC{Int, Int}
+
+    # source and destination of the edge. the index of these arrays are the
+    # edge ids.
+    edge_src::Array{Int, 1}
```
Comment (pjrule) [inline src/graph.jl]: I like how you separated these.

File: src/graph.jl [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    edge_dst::Array{Int, 1}
+
+    # an Array of Arrays where the neighbors of node `i` are found in an array
+    # at index `i`
+    node_neighbors::Array{Array{Int64,1},1}
```
Comment (pjrule) [inline src/graph.jl]: -> `neighbors`?

File: src/graph.jl [hunk 5]
```diff
@@ -0,0 +1,74 @@
+using JSON
+using SparseArrays
+using LightGraphs
+abstract type AbstractGraph end
+
+struct Graph<:AbstractGraph
+    num_nodes::Int
```
Comment (pjrule) [inline src/graph.jl]: -> `n_nodes`? _(discuss)_

File: src/graph.jl [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    num_edges::Int
```
Comment (pjrule) [inline src/graph.jl]: -> `n_edges`? _(discuss)_

File: src/graph.jl [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    # the base SimpleGraph, if we need it
+    simple_graph::SimpleGraph
+end
+
+function Graph(raw_graph::Dict{String, Any},
+               nodes_str::AbstractString = "nodes",
+               adjacency_str::AbstractString = "adjacency",
+               edge_id_str::AbstractString = "id")::Graph
+
+    """ Builds the base Graph object. This is the underlying network of our
```
Comment (pjrule) [inline src/graph.jl]: Rather than "never _need_ to be changed," implying that it _could_ be changed but we prefer not to do so, we should probably just say that `Graph` is immutable.

File: src/graph.jl [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        districts, and should never need to be changed.
+
+        Arguments:
+            filepath: file path to the .json file that contains the graph. This
+                      file is expected to be generated by the `Graph.to_json()`
+                      function http://tiny.cc/7ya2jz of the Python implementation
```
Comment (pjrule) [inline src/graph.jl]: Not sure about the `tiny.cc` thing here, as it obscures that it's a reference to the GerryChain documentation... though there's not really a good solution for handling long URLs in docstrings. :\

File: src/partition.jl [hunk 9]
```diff
@@ -0,0 +1,150 @@
+mutable struct Partition
```
Comment (pjrule) [inline src/partition.jl]: shorten names? _(discuss)_

File: src/partition.jl [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+    # get attributes
+    attributes = get_attributes(raw_graph, graph.num_nodes, nodes_str)
+
+    return Partition(num_districts, num_cut_edges, total_population,
+                     district_populations, cut_edges, populations,
+                     assignments, district_adj, district_nodes, attributes)
+end
+
+################################################################################
```
Comment (pjrule) [inline src/partition.jl]: Just a personal preference, but I've always disliked separators like this—they make the code uglier without providing much semantic value. (I know these decorations are favored by some of the CS faculty here, though 😉)

File: src/partition.jl [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+                                         pop_col::AbstractString,
+                                         assignment_col::AbstractString,
+                                         num_nodes::Int,
+                                         nodes_str::AbstractString = "nodes")
+    """ Returns the arrays of populations and assignments of the graph, where
+        i'th node's population is populations[i] and assignment is assignments[i].
+    """
+    populations = zeros(Int, num_nodes)
+    assignments = zeros(Int, num_nodes)
+    for i = 1:num_nodes
```
Comment (pjrule) [inline src/partition.jl]: I _think_ the Julia idiom here is to use the `in` keyword instead of an assignment operator...

File: src/partition.jl [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        populations[i] = raw_graph[nodes_str][i][pop_col]
+        if raw_graph[nodes_str][i][assignment_col] isa String
```
Comment (pjrule) [inline src/partition.jl]: Cool. This gets around some of the issues with iffy typing that crop up in the shapefile -> NetworkX graph -> JSON conversion...

File: src/partition.jl [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    attributes = get_attributes(raw_graph, graph.num_nodes, nodes_str)
+
+    return Partition(num_districts, num_cut_edges, total_population,
+                     district_populations, cut_edges, populations,
+                     assignments, district_adj, district_nodes, attributes)
+end
+
+################################################################################
+################################################################################
+function get_populations_and_assignments(raw_graph::Dict{String, Any},
```
Comment (pjrule) [inline src/partition.jl]: Including `get_` in these function names may be unnecessary.

**Comment by pjrule** [comment]
Re: `BitSet`—it's an implementation of `Set` optimized specifically for integers. It yields notable performance improvements over `Set` in some cases but has roughly the same interface.

File: src/graph.jl [hunk 14]
```diff
@@ -0,0 +1,74 @@
+using JSON
+using SparseArrays
+using LightGraphs
+abstract type AbstractGraph end
+
+struct Graph<:AbstractGraph
```
Comment (pjrule) [inline src/graph.jl]: Readability: This struct definition should look more like the `Partition` definition (which is formatted beautifully!) if possible. Too much spacing makes it hard to parse imo.

File: src/partition.jl [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    district_nodes::Array{Set{Int}, 1}
+    attributes::Array{Dict{String, Any}}
+end
+
+function Partition(graph::Graph,
+                   raw_graph::Dict{String, Any},
+                   pop_col::AbstractString,
+                   assignment_col::AbstractString,
+                   nodes_str::AbstractString = "nodes")::Partition
+    """ Builds the Partition Struct, which stores all the information of our
```
Comment (pjrule) [inline src/partition.jl]: The discussion of secret-keeping here is informative and probably necessary, but we should think about what tone we want our documentation to take—let's take a look at the GerryChain documentation (which is quite good!) together and come up with some loose guidelines?

File: src/utils.jl [hunk 16]
```diff
@@ -0,0 +1,14 @@
+# general utility functions
```
Comment (pjrule) [inline src/utils.jl]: hmm... do we want to encourage this pattern?

File: test/partition.jl [hunk 17]
```diff
@@ -0,0 +1,95 @@
+# Partition tests
```
Comment (pjrule) [inline test/partition.jl]: I trust that these pass. Let's set up CI/CD ASAP!

File: test/partition.jl (hunk 1 above)
Comment (bsuwal) [reply test/partition.jl]: ok will do

File: test/graph.jl (hunk 2 above)
Comment (bsuwal) [reply test/graph.jl]: will do

File: test/partition.jl (hunk 17 above)
Comment (bsuwal) [reply test/partition.jl]: will do

File: src/partition.jl [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    num_cut_edges::Int
+    total_population::Int
+
+    district_populations::Array{Int, 1} # of length(num_districts)
+    cut_edges::Array{Int, 1}            # of length(num_edges)
+    populations::Array{Int, 1}          # of length(num_nodes)
+    assignments::Array{Int, 1}          # of length(num_nodes)
+    district_adj::SparseMatrixCSC{Int, Int}
+    district_nodes::Array{Set{Int}, 1}
+    attributes::Array{Dict{String, Any}}
```
Comment (pjrule) [inline src/partition.jl]: Discussed at meeting: move `attributes`, `total_population`, and `populations` to `Graph` object to avoid copying over and over!

**Comment by pjrule** [comment]
Abbreviations we decided on:
* `num` stays as `num` (not shortened to `n`)
* `population` -> `pop`
* `district` -> `dist` when appropriate
* `total` stays as `total` (not shortened to `tot`)

File: src/graph.jl (hunk 4 above)
Comment (bsuwal) [reply src/graph.jl]: ok: node_neighbors -> neighbors, edge_src -> src, edge_dst-> dst

File: src/graph.jl (hunk 7 above)
Comment (bsuwal) [reply src/graph.jl]: ok.

File: src/graph.jl (hunk 8 above)
Comment (bsuwal) [reply src/graph.jl]: leave a citation for the url instead of the tiny cc

File: src/partition.jl (hunk 15 above)
Comment (bsuwal) [reply src/partition.jl]: A discussion of data structures is not necessary - user-specific docstrings would be more helpful.

File: src/partition.jl (hunk 10 above)
Comment (bsuwal) [reply src/partition.jl]: can you tell I went to Tufts CS?

Comment (bsuwal) [reply src/partition.jl]: ill chuck it out

File: src/partition.jl (hunk 11 above)
Comment (bsuwal) [reply src/partition.jl]: for i in 1:num_nodes

Cool!

File: src/utils.jl (hunk 16 above)
Comment (bsuwal) [reply src/utils.jl]: ... lets get back to this after the Graph-Partition refactor

File: src/graph.jl (hunk 4 above)
Comment (bsuwal) [reply src/graph.jl]: running into a problem: the lightgraphs library has a function called src() and dst(), and this is causing namespace issues. In the Graph function we want to use the src()/dst() functions, and multiple dispatch doesnt work on data structures i dont think. Leaving it as edge_src for now but I agree this is kinda ugly. Will think more.

File: src/graph.jl (hunk 3 above)
Comment (bsuwal) [reply src/graph.jl]: thanks!

File: src/graph.jl (hunk 8 above)
Comment (bsuwal) [reply src/graph.jl]: The citation thing turns out to still be rather ugly... but I am fine leaving the long url here.

File: src/utils.jl (hunk 16 above)
Comment (bsuwal) [reply src/utils.jl]: Maybe we should throw away utils.jl, for now atleast. Currently after the refactor Graph and Partition both still take in the filepath, which I still think is a bit messy, but I am willing to live with it........ for now! Will keep thinking about this
